#! /bin/sh /usr/share/dpatch/dpatch-run
## CVE-2008-2392.dpatch by  <andrea.de.iacovo@gmail.com>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Disables unfiltered upload capability for administrators.

@DPATCH@

diff -Nru wordpress-old/wp-admin/includes/schema.php wordpress-2.7.1/wp-admin/includes/schema.php
--- wordpress-old/wp-admin/includes/schema.php	2009-02-12 09:13:33.000000000 +0100
+++ wordpress-2.7.1/wp-admin/includes/schema.php	2009-02-12 09:13:44.000000000 +0100
@@ -482,11 +482,11 @@
  * @since 2.3.0
  */
 function populate_roles_230() {
-	$role =& get_role( 'administrator' );
-
-	if ( !empty( $role ) ) {
-		$role->add_cap( 'unfiltered_upload' );
-	}
+#	$role =& get_role( 'administrator' );
+#
+#	if ( !empty( $role ) ) {
+#		$role->add_cap( 'unfiltered_upload' );
+#	}
 }
 
 /**
diff -Nru wordpress-old/wp-admin/menu.php wordpress-2.7.1/wp-admin/menu.php
--- wordpress-old/wp-admin/menu.php	2009-01-22 19:56:10.000000000 +0100
+++ wordpress-2.7.1/wp-admin/menu.php	2009-02-12 08:37:00.000000000 +0100
@@ -98,6 +98,7 @@
 	$submenu['options-general.php'][35] = array(__('Privacy'), 'manage_options', 'options-privacy.php');
 	$submenu['options-general.php'][40] = array(__('Permalinks'), 'manage_options', 'options-permalink.php');
 	$submenu['options-general.php'][45] = array(__('Miscellaneous'), 'manage_options', 'options-misc.php');
+	$submenu['options-general.php'][45] = array(__('Security'), 'manage_options', 'options-security.php');
 
 $_wp_last_utility_menu = 60; // The index of the last top-level menu in the utility menu group
 
diff -Nru wordpress-old/wp-admin/options-sec.php wordpress-2.7.1/wp-admin/options-sec.php
--- wordpress-old/wp-admin/options-sec.php	1970-01-01 01:00:00.000000000 +0100
+++ wordpress-2.7.1/wp-admin/options-sec.php	2009-02-12 09:10:48.000000000 +0100
@@ -0,0 +1,48 @@
+<?php
+/**
+ * Options Management Administration Panel.
+ *
+ * Just allows for displaying of options.
+ *
+ * This isn't referenced or linked to, but will show all of the options and
+ * allow editing. The issue is that serialized data is not supported to be
+ * modified. Options can not be removed.
+ *
+ * @package WordPress
+ * @subpackage Administration
+ */
+
+/** WordPress Administration Bootstrap */
+require_once('admin.php');
+
+$title = __('Security Settings');
+$this_file = 'options-sec.php';
+$parent_file = 'options-security.php';
+
+if ( !current_user_can('manage_options') )
+	wp_die(__('Cheatin&#8217; uh?'));
+
+$role = get_role('administrator');
+$change_ok =  false;
+
+if(empty($role))
+	wp_die(__("Something went wrong. Please try again"));
+
+if (!isset($_POST["unfiltered_upload"]) && $_POST["unf_up"]=="true"){
+	$role->remove_cap('unfiltered_upload');
+	$change_ok = true;
+	}
+else if (isset($_POST["unfiltered_upload"]) && $_POST["unf_up"]=="false"){
+	$role->add_cap('unfiltered_upload');
+	$change_ok = true;
+	}
+	
+$wp_rewrite->flush_rules();
+
+wp_cache_flush();
+
+$goback = add_query_arg( 'updated', $change_ok, wp_get_referer() );
+wp_redirect( $goback );
+
+include('admin-footer.php');
+?>
diff -Nru wordpress-old/wp-admin/options-security.php wordpress-2.7.1/wp-admin/options-security.php
--- wordpress-old/wp-admin/options-security.php	1970-01-01 01:00:00.000000000 +0100
+++ wordpress-2.7.1/wp-admin/options-security.php	2009-02-12 08:58:59.000000000 +0100
@@ -0,0 +1,81 @@
+<?php
+/**
+ * General settings administration panel.
+ *
+ * @package WordPress
+ * @subpackage Administration
+ */
+
+/** WordPress Administration Bootstrap */
+require_once('./admin.php');
+
+$title = __('Security Settings');
+$parent_file = 'options-security.php';
+
+/**
+ * Display JavaScript on the page.
+ *
+ * @package WordPress
+ * @subpackage General_Settings_Panel
+ */
+function add_js() {
+?>
+<script type="text/javascript">
+//<![CDATA[
+	jQuery(document).ready(function($){
+		$("input[name='date_format']").click(function(){
+			if ( "date_format_custom_radio" != $(this).attr("id") )
+				$("input[name='date_format_custom']").val( $(this).val() );
+		});
+		$("input[name='date_format_custom']").focus(function(){
+			$("#date_format_custom_radio").attr("checked", "checked");
+		});
+
+		$("input[name='time_format']").click(function(){
+			if ( "time_format_custom_radio" != $(this).attr("id") )
+				$("input[name='time_format_custom']").val( $(this).val() );
+		});
+		$("input[name='time_format_custom']").focus(function(){
+			$("#time_format_custom_radio").attr("checked", "checked");
+		});
+	});
+//]]>
+</script>
+<?php
+}
+add_filter('admin_head', 'add_js');
+
+include('./admin-header.php');
+?>
+
+<div class="wrap">
+<?php screen_icon(); ?>
+<h2><?php echo wp_specialchars( $title ); ?></h2>
+
+<form method="post" action="options-sec.php">
+<?php settings_fields('general'); ?>
+
+<table class="form-table">
+<tr valign="top">
+<th scope="row"><?php _e('Unfiltered upload') ?></th>
+<td> <fieldset><legend class="hidden"><?php _e('Unfiltered upload') ?></legend><label for="unfiltered_upload">
+<input name="unfiltered_upload" type="checkbox" id="unfiltered_upload" value="1" <?php checked('1', get_role('administrator')->has_cap('unfiltered_upload')); ?> />
+<?php _e('Admins can upload any type of file') ?></label>
+<?php
+$unf_up = get_role("administrator")->has_cap("unfiltered_upload");
+if (!$unf_up)
+	print('<input name="unf_up" type="hidden" value="false">');
+else
+	print('<input name="unf_up" type="hidden" value="true">');
+?>
+</fieldset></td>
+</tr>
+</table>
+<p class="submit">
+<input type="submit" name="Submit" class="button-primary" value="<?php _e('Save Changes') ?>" />
+</p>
+</form>
+
+</div>
+
+<?php include('./admin-footer.php') ?>
